sqlglot/optimizer/qualify_columns.py

```diff
@@ -296,23 +297,29 @@ def replace_columns(
             if expand_only_groupby and is_group_by and column.parent is not node:
                 continue
 
+            skip_replace = False
             table = resolver.get_table(column.name) if resolve_table and not column.table else None
             alias_expr, i = alias_to_expression.get(column.name, (None, 1))
-            double_agg = (
-                (
+
+            if alias_expr:
+                skip_replace = bool(
                     alias_expr.find(exp.AggFunc)
-                    and (
-                        column.find_ancestor(exp.AggFunc)
-                        and not isinstance(column.find_ancestor(exp.Window, exp.Select), exp.Window)
-                    )
+                    and column.find_ancestor(exp.AggFunc)
+                    and not isinstance(column.find_ancestor(exp.Window, exp.Select), exp.Window)
                 )
-                if alias_expr
-                else False
-            )
 
-            if table and (not alias_expr or double_agg):
+                # BigQuery's having clause gets confused if an alias matches a source.
+                # SELECT x.a, max(x.b) as x FROM x GROUP BY 1 HAVING x > 1;
+                # If HAVING x is expanded to max(x.b), bigquery treats x as the new projection x instead of the table
+                if is_having and dialect == "bigquery":
+                    skip_replace = skip_replace or any(
+                        node.parts[0].name in projections
+                        for node in alias_expr.find_all(exp.Column)
+                    )
```

Comment:
- Should this be `find_all_in_scope`?
- Could this ast traversal be factored out maybe? Or cached so we don't recompute this on each iteration. E.g. the outer loop already checks columns:

```python
        for column in walk_in_scope(node, prune=lambda node: node.is_star):
            if not isinstance(column, exp.Column):
                continue
```